lib/Server.js

```diff
@@ -201,349 +198,22 @@ class Server {
-          app.get('*', (req, res) => {
-            res.writeHead(302, {
-              Location: contentBase + req.path + (req._parsedUrl.search || ''),
-            });
-
-            res.end();
-          });
-        } else if (typeof contentBase === 'number') {
-          this.log.warn(
-            'Using a number as contentBase is deprecated and will be removed in the next major version. Please use the proxy option instead.'
-          );
-
-          this.log.warn(
-            'proxy: {\n\t"*": "//localhost:<your current contentBase configuration>"\n}'
-          );
-          // Redirect every request to the port contentBase
-          app.get('*', (req, res) => {
-            res.writeHead(302, {
-              Location: `//localhost:${contentBase}${req.path}${req._parsedUrl
-                .search || ''}`,
-            });
-
-            res.end();
-          });
-        } else {
-          // route content request
-          app.get('*', express.static(contentBase, options.staticOptions));
-        }
-      },
-      contentBaseIndex: () => {
-        if (Array.isArray(contentBase)) {
-          contentBase.forEach((item) => {
-            app.get('*', serveIndex(item));
-          });
-        } else if (
-          !/^(https?:)?\/\//.test(contentBase) &&
-          typeof contentBase !== 'number'
-        ) {
-          app.get('*', serveIndex(contentBase));
-        }
-      },
-      watchContentBase: () => {
-        if (
-          /^(https?:)?\/\//.test(contentBase) ||
-          typeof contentBase === 'number'
-        ) {
-          throw new Error('Watching remote files is not supported.');
-        } else if (Array.isArray(contentBase)) {
-          contentBase.forEach((item) => {
-            this._watch(item);
-          });
-        } else {
-          this._watch(contentBase);
-        }
-      },
-      before: () => {
-        if (typeof options.before === 'function') {
-          options.before(app, this, compiler);
-        }
-      },
-      middleware: () => {
-        // include our middleware to ensure
-        // it is able to handle '/index.html' request after redirect
-        app.use(this.middleware);
-      },
-      after: () => {
-        if (typeof options.after === 'function') {
-          options.after(app, this, compiler);
-        }
-      },
-      headers: () => {
-        app.all('*', this.setContentHeaders.bind(this));
-      },
-      magicHtml: () => {
-        app.get('*', this.serveMagicHtml.bind(this));
-      },
-      setup: () => {
-        if (typeof options.setup === 'function') {
-          this.log.warn(
-            'The `setup` option is deprecated and will be removed in v4. Please update your config to use `before`'
-          );
-
-          options.setup(app, this);
-        }
-      },
-    };
-
+    const features = createFeatures.call(this, {
+      app,
+      compiler,
+      options,
+      contentBase,
+      websocketProxies,
+    });
```

Comment:
Maybe better do this as method `setupFeatures` (we need `this` in context so better use SOLID practice here and do this as method)

Reply:
Apparently I didn’t explain well :smile: I meant to do method `Server.setupFeatures` like:
```js
class Server {
  setupFeatures() {
    // Logic here
  }
}
```

Reply:
like this?

```javascript
    const features = this.setupFeatures({
      app,
      compiler,
      options,
      contentBase,
      websocketProxies,
    });

...

  setupFeatures(...args) {
    return createFeatures.call(this, args);
  }

```

Reply:
put code from `setupFeatures` inside method (SOLID), we use this and using `call` with `this` is not good idea

Reply:
hmm... but `createFeatures` has more than 280 lines... and L591 is already used  `call`.

Reply:
ok, fixed https://github.com/webpack/webpack-dev-server/pull/1839/commits/1ea70ce11d777388895cba111e51ee59f3d607c1